GraphBase/src/graph/Graph.java

```diff
@@ -0,0 +1,116 @@
+package graph;
+
+import java.util.Map;
+import java.util.HashMap;
+import java.util.NoSuchElementException;
+
+public abstract class Graph {
+
+  protected Map<Integer, Vertex> vertices;
+
+  public Graph() {
+    this.vertices = new HashMap<Integer, Vertex>();
+  }
+
+  public void addVertex(int newVertexId) {
+    Vertex newVertex = new Vertex(newVertexId);
+    vertices.put(newVertex.getId(), newVertex);
+  }
+
+  public void addVertex(Vertex newVertex) {
+    if (vertices.keySet().contains(newVertex.getId())) {
+      System.out.println("The vertex has already been exsited");
+    }
+  }
+
+  public void deleteVertex(int vertexId) {
+    if (!vertices.containsKey(vertexId)) {
+      throw new NoSuchElementException("The vertex id is not existed");
+    }
+    vertices.remove(vertexId);
+    for (Integer key : vertices.keySet()) {
+      if (vertices.get(key).getAdjEdges().keySet().contains(vertexId)) {
+        vertices.get(key).getAdjEdges().remove(vertexId);
+      }
+    }
+  }
+
+  public void deleteVertex(Vertex vertex) {
+    deleteVertex(vertex.getId());
+  }
+
+  public void addEdge(int fromId, int toId) {
+    if (fromId == toId) {
+      System.out.println("Loop is not allowed");
+      return;
+    }
+    if (!vertices.containsKey(fromId)) {
+      addVertex(fromId);
+    }
+    if (!vertices.containsKey(toId)) {
+      addVertex(toId);
+    }
+    if (!this.vertices.get(fromId).isAdjacent(toId)) {
+      Edge newEdge = new Edge(fromId, toId);
+      this.vertices.get(fromId).addAdjEdge(newEdge);
+    }
+  }
+
+  public void addEdge(Vertex fromVertex, Vertex toVertex) {
+    if (fromVertex.getId() == toVertex.getId()) {
```

Comment:
I do not think you should implement addEdge here. The adding edge behaviour is different for directed and undirected graph. Also, what if you have graph that supports loop and multi-edges?

Reply:
Leave it empty and define in corresponding extended graph accordingly. And do so for all the following edge operations.

Reply:
Is it proper to use weight represents multiedges.

Reply:
Because use HashMap to store the adjEdges, and the id I use is the same id with the edge's to Vertex

Reply:
What I'm saying is. addEdge should be repectively implemented in UndirectedGraph, DirectedGraph, and probably LoopGraph or MultiEdgeGraph? They all have different behaviour. As a result, do not implement it in base graph and just leave it empty. The good things of leaving it empty in Base Graph is that every inheritance should implement it, which enforces you to control their behaviour.